File: 147 Insertion Sort List/147.java

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode insertionSortList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        // initial a dummy node , the dummy.next to represent the start of the sorted list
        // because during the sort, the head can not be definitely determined
        ListNode dummy = new ListNode(0);
        // the start of the node that prepare to be sorted
        ListNode cur = head;

        while (cur != null) {
            ListNode nextNode = cur.next;
           
            ListNode p = dummy;
            while (p.next != null && p.next.val < cur.val) {
                p = p.next;
            }

            cur.next = p.next;
            p.next = cur;
            cur = nextNode;
        }

        return dummy.next;

    }

}

/*
  Time complexity O(n^2)
  Space complexity O(1)
*/

File: 147 Insertion Sort List/147.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var insertionSortList = function(head) {
    if (!head || !head.next) {
        return head;
    }

    let cur = head
    let dummy = new ListNode(0)
    while (cur) {
        let nextCur = cur.next
        let p = dummy
        while (p.next && p.next.val < cur.val) {
            p = p.next
        }
        let nextNode = p.next
        p.next = cur
        cur.next = nextNode
        cur = nextCur
    }

    return dummy.next;
};